### PR TITLE
Shortestpath

### PR DESCRIPTION
**Pull request by jackwener**
## What type of PR is this?
- [ ] bug
- [x] feature
- [ ] enhancement

## Checklist:
Tests:
- [ ] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

File: src/graph/executor/query/ShortestPathExecutor.cpp [hunk 1]
```diff
@@ -0,0 +1,319 @@
+/* Copyright (c) 2021 vesoft inc. All rights reserved.
```
Comment (darionyaphet) [inline src/graph/executor/query/ShortestPathExecutor.cpp:1]: 2022?

File: src/graph/executor/query/ShortestPathExecutor.cpp [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+    step_ = 0;
+    do {
+      // Set left(dss_[0]) or right(dss_[1]) dataset as req dataset, and get neighbors
+      for (int i = 0; i < 2; i++) {
+        if (break_ || step_ > shortestPathNode_->stepRange()->max()) {
+          dss_[0].clear();
+          dss_[1].clear();
+          break;
+        }
+        status = getNeighbors().get();
```
[the remark this answers is not in the document]
Comment (CPWstatic) [reply src/graph/executor/query/ShortestPathExecutor.cpp:49]: Expansion in both directions should be concurrent.

File: .linters/cpp/checkKeyword.py [hunk 3]
```diff
@@ -119,6 +119,8 @@
     'KW_GEOGRAPHY',
     'KW_DURATION',
     'KW_ACROSS',
+    "KW_SHORTESTPATH",
```
Comment (Shylock-Hg) [inline .linters/cpp/checkKeyword.py:122]: New reserved word means incompatible.

File: src/graph/executor/query/ShortestPathExecutor.cpp [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+  vidToVertex_.clear();
+  allPrevPaths_[0].clear();
+  allPrevPaths_[1].clear();
+
+  return Executor::close();
+}
+
+Status ShortestPathExecutor::buildRequestDataSet() {
+  auto inputVar = shortestPathNode_->inputVar();
+  const Result& inputResult = ectx_->getResult(inputVar);
```
Comment (nevermore3) [inline src/graph/executor/query/ShortestPathExecutor.cpp:91]: You can do the Cartesian product directly without the Cartesian product operator

File: src/graph/executor/query/ShortestPathExecutor.cpp [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+    }
+    if (eFilter != nullptr) {
+      auto& eFilterVal = eFilter->eval(ctx(iter));
+      if (!eFilterVal.isBool() || !eFilterVal.getBool()) {
+        continue;
+      }
+    }
+    Value srcV = iter->getVertex();
+    Value edge = iter->getEdge();
+    const Value& dst = iter->getEdgeProp("*", kDst);
```
Comment (nevermore3) [inline src/graph/executor/query/ShortestPathExecutor.cpp:207]: dst = edge.dst

File: src/graph/executor/query/ShortestPathExecutor.cpp [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+    if (!SchemaUtil::isValidVid(dst, *(spaceInfo.spaceDesc.vid_type_ref()))) {
+      continue;
+    }
+
+    // Due to src can have multiple edge, so src will be repeated, this make it work once
```
Comment (nevermore3) [inline src/graph/executor/query/ShortestPathExecutor.cpp:212]: vidToVertex_.find(edge.src)

File: src/graph/executor/query/ShortestPathExecutor.h [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+      allPrevPaths_.emplace_back();
+    }
+  }
+
+  folly::Future<Status> execute() override;
+
+  Status close() override;
+
+ private:
+  using Paths = std::vector<Row>;
```
Comment (nevermore3) [inline src/graph/executor/query/ShortestPathExecutor.h:43]: Rows

**Comment by Sophie-Xie** [comment]
@nevermore3 will create a new PR

**Pull request closed without merging**